src/v/raft/compaction_coordinator.cc

```diff
@@ -0,0 +1,296 @@
+            // leader change in order to avoid RPC storm on startup
+            _has_seen_a_leader = true;
+            _need_force_update = true;
+            return;
+        }
+        if (_leader_id != _self) {
+            send_local_mcco_to_leader();
+        } else {
+            recalculate_mtro();
+            on_mtro_update();
+        }
+    }
+}
+
+void compaction_coordinator::on_group_configuration_change() {
+    if (!_started) {
+        return;
+    }
+    if (is_leader()) {
+        _need_force_update = true;
+        recalculate_mtro();
+    }
+}
+
+bool compaction_coordinator::process_remote_max_cleanly_compacted_offset(
+  vnode node_id, model::offset remote_mcco) {
+    if (!_started) {
+        return false;
+    }
+    vlog(
+      _logger.debug, "received MCCO: {} from vnode: {}", remote_mcco, node_id);
+    if (!is_leader()) {
+        vlog(
+          _logger.debug,
+          "received MCCO from vnode: {}, but we're not the leader",
+          node_id);
+        return false;
+    }
+    auto fstat_it = _fstats.find(node_id);
+    if (fstat_it == _fstats.end()) {
+        vlog(
+          _logger.debug,
+          "received MCCO from vnode: {}, who is not in the configuration",
+          node_id);
+        return false;
+    }
+    auto& stat_mcco = fstat_it->second.max_cleanly_compacted_offset;
+    // just in case of out-of-order messages
+    stat_mcco = std::max(stat_mcco, remote_mcco);
+    recalculate_mtro();
+    return true;
+}
+
+model::offset compaction_coordinator::get_max_tombstone_remove_offset() const {
+    return _mtro;
+}
+
+model::offset
+compaction_coordinator::get_local_max_cleanly_compacted_offset() const {
+    return _local_mcco;
+}
+
+compaction_coordinator::clock_t::duration
+compaction_coordinator::base_interval() const {
+    auto retention = _log->config().tombstone_retention_ms();
+    // Typically retention is 24h, set update interval to 1h
+    // (up to 1.5h due to jitter).
+    // Using half-max for disabled to avoid overflow in jitter.
+    clock_t::duration base_interval = retention
+                                        ? clock_t::duration(*retention / 24)
+                                        : clock_t::duration::max() / 2;
+    vlog(_logger.debug, "calculated base_interval as {}", base_interval);
+    return base_interval;
+}
+
+void compaction_coordinator::on_ntp_config_change() {
+    auto base = base_interval();
+    _local_mcco_query_timer_jitter = jitter_t{base};
+    _rpc_sender.update_retry_interval(base);
+    // rearm
+    if (_local_mcco_query_timer.armed()) {
+        _local_mcco_query_timer.cancel();
+        arm_timer();
+    }
+}
+
+void compaction_coordinator::update_local_mcco() {
+    on_local_mcco_update(_log->cleanly_compacted_prefix_offset());
+    arm_timer();
+}
+
+void compaction_coordinator::on_local_mcco_update(model::offset new_mcco) {
+    vlog(
+      _logger.debug,
+      "updating local MCCO from {} to {}",
```

Comment:
i am not fan of the acronyms in log messages, can we always use full name in log lines ?